File: 백준/Gold/17178. 줄서기/줄서기.java
import java.util.*;

class Human implements Comparable<Human> {
    String alpha;
    int num;
    
    Human(String alpha, int num) {
        this.alpha = alpha;
        this.num = num;
    }

    @Override
    public int compareTo(Human other) {
        if (this.alpha.equals(other.alpha)) {
            return Integer.compare(this.num, other.num);
        }
        return this.alpha.compareTo(other.alpha);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Human other = (Human) obj;
        return this.alpha.equals(other.alpha) && this.num == other.num;
    }

    @Override
    public int hashCode() {
        return Objects.hash(alpha, num);
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int N = scanner.nextInt();
        scanner.nextLine();

        List<Human> lineList = new ArrayList<>();
        List<Human> lineSortList = new ArrayList<>();
        Stack<Human> waitStk = new Stack<>();

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < 5; j++) {
                String s = scanner.next();
                String alpha = s.substring(0, 1);
                int num = Integer.parseInt(s.substring(2));
                Human human = new Human(alpha, num);
                lineList.add(human);
                lineSortList.add(human);
            }
        }

        Collections.sort(lineSortList);
        int curIdx = 0;

        for (int i = 0; i < N * 5; i++) {
            if (lineSortList.get(curIdx).equals(lineList.get(i))) {
                curIdx++;
            } else if (!waitStk.isEmpty() && waitStk.peek().equals(lineSortList.get(curIdx))) {
                curIdx++;
                waitStk.pop();
                i--;
            } else {
                waitStk.push(lineList.get(i));
            }
        }

        boolean flag = true;

        while (!waitStk.isEmpty()) {
            if (waitStk.peek().equals(lineSortList.get(curIdx))) {
                curIdx++;
                waitStk.pop();
            } else {
                flag = false;
                break;
            }
        }
        System.out.println(flag ? "GOOD" : "BAD");
    }
}

File: 백준/Gold/17178. 줄서기/줄서기.cc
#include<iostream>
#include<algorithm>
#include<vector>
#include<string>
#include<stack>
using namespace std;

int N;
string s;

class human {
public: 
	string alpha;
	int num;


	human(string alpha, int num) {
		this->alpha = alpha;
		this->num = num;
	}

	bool operator<(const human& other) const {
		if (alpha == other.alpha) {
			return num < other.num;
		}
		return alpha < other.alpha;
	}

	bool operator==(const human& other) const {
		return alpha == other.alpha && num == other.num;
	}

};

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	vector<human>lineList;
	vector<human>lineSortList;
	stack<human>waitStk;

	cin >> N;

	for (int i = 0; i < N; i++) {
		for (int j = 0; j < 5; j++) {
			cin >> s;
			string alpha = s.substr(0, 1);
			int num = stoi(s.substr(2));
			lineList.push_back(human(alpha, num));
			lineSortList.push_back(human(alpha, num));
		}
	}

	sort(lineSortList.begin(), lineSortList.end());
	int curIdx = 0;

	for (int i = 0; i < N * 5; i++) {

		// 넘어가진다면? 
		if (lineSortList[curIdx] == lineList[i]) {
			curIdx++;
		}
		else if (!waitStk.empty() && waitStk.top() == lineSortList[curIdx]) {
			// 스택에서 빠진다면?
			curIdx++;
			waitStk.pop();
			i--;
		}
		else {
			waitStk.push(lineList[i]);
		}
	}
	bool flag = true;
	
	while (waitStk.size()) {

		if (waitStk.top() == lineSortList[curIdx]) {
			curIdx++;
			waitStk.pop();
		}
		else {
			flag = false;
			break;
		}
	}

	string result = "GOOD";
	if (!flag) result = "BAD";

	cout << result;

	return 0;
}